src/Widgets/CameraView.vala

```
@@ -17,77 +17,201 @@
-        play_shutter_sound ();
+        var spinner = new Gtk.Spinner ();
+        spinner.active = true;
 
-        return true;
-    }
+        status_label = new Gtk.Label (null);
+
+        status_grid = new Gtk.Grid ();
+        status_grid.row_spacing = 6;
+        status_grid.orientation = Gtk.Orientation.VERTICAL;
+        status_grid.halign = Gtk.Align.CENTER;
+        status_grid.valign = Gtk.Align.CENTER;
+        status_grid.add (spinner);
+        status_grid.add (status_label);
+        add_named (status_grid, "status");
 
-    public bool start_recording () {
-        if (!this.is_ready_for_capture () || this.is_recording_video ()) {
-            warning ("Device isn't ready for recording videos.");
+        no_device_view = new Granite.Widgets.AlertView (_("No Supported Camera Found"),
+                                                        _("Connect a webcam or other supported video device to take photos and video."),
+                                                        "accessories-camera");
+        add_named (no_device_view, "no-device");
 
-            return false;
+        if (infos.length == 0) {
+            visible_child = no_device_view;
+            return;
         }
 
-        this.start_video_recording (Utils.get_new_media_filename (Utils.ActionType.VIDEO));
+        try {
+            pipeline = Gst.parse_launch ("v4l2src name=v4l2src ! videoflip method=horizontal-flip ! tee name=tee ! queue leaky=downstream max-size-buffers=10 ! videoconvert ! videoscale ! gtksink name=gtksink") as Gst.Pipeline;
+            v4l2src = pipeline.get_by_name ("v4l2src");
+            tee = pipeline.get_by_name ("tee");
+            var gtksink = pipeline.get_by_name ("gtksink");
+            gtksink.get ("widget", out video_widget);
+            video_widget.expand = true;
+            add_named (video_widget, "video");
+            visible_child = status_grid;
+        } catch (Error e) {
+            critical (e.message);
```

Perhaps we should throw a messagedialog here about missing runtime dependencies? That would probably be more friendly for our cross-distro story

---

Done :)